Add docker-ng state module

terminalmage opened the pull request:
This is just the first of several states that will be written.

@ticosax, I'd love any feedback you're able to provide. I'll be adding to this
module over the next week or so, please let me know what kind of states you
think would be useful.

terminalmage committed:
More bugfixes discovered while writing docker-ng states

terminalmage committed:
Docstring tweaks

terminalmage committed:
Add docker-ng state module

This is just the first of several states that will be written.

terminalmage committed:
Add docker-ng state docs to toctree

terminalmage committed:
Fix typo

terminalmage commented:
Keep in mind, mod_watch support hasn't been added yet. It's coming, though.

ticosax commented on salt/states/dockerng.py:
```diff
@@ -0,0 +1,1495 @@
+            ret['comment'] = _format_comments(comments)
+            return ret
+
+        time.sleep(2)
```
is it really necessary ?

terminalmage replied:
Good catch! That was in there for testing purposes when I was developing, and I forgot to remove it. I'll get it when I make the changes I mentioned [here](https://github.com/saltstack/salt/pull/22548#issuecomment-91996517).

ticosax commented:
@terminalmage 
You did an amazing job. It is really neat. I do not see anything preventing me to serve my use cases with `docker-ng`.
I will give it a try as soon as I can, but it won't be in next days. Unfortunately I need to do other things before.
When you will provide `mod_watch` please can you consider back-porting 4c2358150593622ce409c6f749b33dadf7e059e2 ?
Otherwise big :+1:

terminalmage commented:
@ticosax thanks! No worries about the wait, I expect to be adding more states anyway.

I will need to review the commit you posted on Monday, but I'm pretty sure we can get it in.

As for the one state that I have already written, I think that I am going to uncouple the parts that handle pulling/building/loading an image into a separate state. If you have 10 containers that rely on the same image, it's much easier to have one state handling the image, and then use requisites to signal the other states that the containers need to re-created. Not only that, but you'll have one state pulling the image instead of 10. One other problem with the state (as it is currently constructed) is that if you have more than one container using the same image, and one container ends up updating that image, then when the other containers pull and no changes are detected, those other containers don't get rebuilt. That's no good!

So, I am going to request that we hold off on merging this for now, so that I can make the needed changes.

thatch45 commented:
whats the status here @terminalmage ?

terminalmage commented:
Still working on it, should have this PR updated later today. Other things (packaging, etc.) have kept me from working on this since last week.

thatch45 commented:
ok, just let me know when it is ready! Thanks @terminalmage !

terminalmage committed:
Make salt.utils.clean_kwargs() clean all kwargs starting with dunder

terminalmage committed:
Fixes/additions to dockerng module to work better with states

terminalmage committed:
Refactor/add to dockerng states

terminalmage committed:
Lint fixes

terminalmage committed:
Rename references to docker-ng with dockerng

This should clear up confusion that may be caused by the virtual name
being different from the module name.

terminalmage commented:
Pull request has been updated. The one state that I had added has been refactored, and several more have been added.

@ticosax I was able to get your soft-reload functionality into the ``dockerng.running`` state. See the ``watch_action`` param.

terminalmage commented:
Test failure looks legit, I will investigate.

ticosax commented:
Thank you, @terminalmage 
I hope it will be merged soon. I would like to give a try.
Also, we will need to extend support for salt.modules.mine with `dockerng`. I can help on this matter if it is not your priority.

terminalmage commented:
Test results have been cycled out, triggered the tests to run again to make sure I address all failing tests.

terminalmage committed:
Update unit.utils.utils_test.UtilsTestCase.test_clean_kwargs

This reflects the new behavior in which all dunder kwargs are cleaned

terminalmage commented:
The CentOS test flaked out and didn't send back results, but the Ubuntu test passed. Re-running just to be certain. Lint tests will fail but are unrelated.

ticosax commented:
:sparkles: Next week is gonna be fun. Thank you for this new toy.

terminalmage commented:
@ticosax you're welcome! There are another couple states that I need to add, but this is a good start :)

Note that "docker-ng" has become "dockerng" in all usage. After some internal discussions, we decided that this would be best to reduce confusion, since the module name and the CLI usage would be different.

terminalmage commented:
@ticosax Also, I got your soft reload feature ported in, but since I wanted to be able to trigger a container rebuild via a watch requisite as well, I added an option to the ``dockerng.running`` state called ``watch_action`` which lets you specify a signal to send when a watch requisite is triggered.
